feat: AI block UI

matthewlipski opened the pull request:
![ai](https://github.com/user-attachments/assets/fe32677b-182c-4b80-9dd5-3e184f18b864)


This PR adds all the UX/UI for AI integration in BlockNote according to the [Notion doc](https://www.notion.so/typecell/Spec-1dd474b463fa42098ae40013b244855b). The UI is not yet plugged into an API for querying AI models, so it uses a mock function which accepts a string prompt and returns a string AI response.

There are a few differences in the implementation vs the spec listed in the doc:

- The AI inline toolbar now instantly opens after selecting a prompt. This also means there are no AI related loaders outside of the AI inline toolbar. Currently the loader is just an `Updating...` string but can easily be changed.
- The "Accept", "Retry", and "Revert" buttons shown after an AI response is inserted are an entirely separate toolbar, instead of being part of the formatting toolbar.
    - Inserting content via the AI suggestion menu also shows this separate toolbar.

TODOs:

- [x] AI Block update button text isn't being set to `Updating...` on click.
- [x] Autofocus not working for prompt input in the AI button popover in the formatting toolbar.
- [ ] Error is thrown when calling `mockAIReplaceSelection` or `mockAIInsertAfterSelection` in `AIInlineToolbar` while React strict mode is enabled. Everything works fine for the user though.
- [x] The AI suggestion menu items are currently derived from dictionary entries, which we probably want to change.
- [ ] AI menu close on escape.
- [ ] AI menu close on backspace.
- [ ] AI menu fade in.
- [ ] AI menu smaller height.
- [ ] AI menu plugin should pass previous block pos when current block is empty. Currently this position change is done in the controller via CSS.

matthewlipski committed:
Added AI block

YousefED reviewed (commented):
In general looks good, I only looked at the overall architecture for now (not line-by-line). So far I have a few main comments:

**Extracting package**
Just realizing this now, I think we should try to ship this as a separate package. Two reasons for this: (a) it's likely this may be licensed differently (i.e.: GPL / Pro only) and (b) it would be a good "proof of the pudding" to see whether we can externalize larger plugins / extensions in a way that still makes sense


**Duplicated toolbar code?**
I think we're starting to see a lot of similar code with the AI Toolbar and plugin that we already have for FilePanel and LinkToolbar, right? If so, it might be a sign that we need to generalize this code and reduce the duplicate code

**AI specific**
- We need a way to pass in the `mockAIModelCall` function
- The AI Model call might need access to the rest of the document / blocks above / below the block
- I think the current mockup is some good first steps but the Notion AI Block is definitely more polished. Specifically, I like their "Generated by AI x minutes ago" context and don't really like our "show prompt" button.

wdyt?

matthewlipski committed:
Added inline and slash menu AI

matthewlipski committed:
Small fix

matthewlipski committed:
UX improvements & refactor

matthewlipski committed:
Extracted AI to separate package & changed AI block toolbar UX

matthewlipski committed:
Finished initial package split

matthewlipski committed:
Moved last AI references to AI package

matthewlipski committed:
Reverted minor unneeded changes

YousefED committed:
refactor architecture

YousefED committed:
add extensions

matthewlipski committed:
Refactored AI dictionary

YousefED committed:
clean dictionary

YousefED committed:
fix

YousefED committed:
fix

matthewlipski committed:
Made AI button use suggestion menu components

matthewlipski committed:
Added keyboard navigation to AI button

matthewlipski committed:
Refactored AI button

matthewlipski committed:
Changed AI from suggestion menu to propriety menu

matthewlipski committed:
Minor changes

matthewlipski committed:
Prevented focus swapping on suggestion menu items

matthewlipski committed:
- AI Menu input spans full block width
- AI Menu items look more similar to Notion
- `getDefaultAIMenuItems` uses same pattern as `getDefaultSlashMenuItems`
- AI Block actually sets the `timeGenerated` prop after generating a response
- AI Block only appears in the block type dropdown when the selection is in one
- AI button in Formatting Toolbar now opens the AI Menu normally instead of in a popover

matthewlipski committed:
Fixed AI Menu position for empty blocks

matthewlipski committed:
Made AI block react instead of vanilla

YousefED committed:
fix build

YousefED committed:
schema

YousefED committed:
improve json schema methods

YousefED committed:
Merge remote-tracking branch 'origin/main' into ai-block

YousefED committed:
merge

YousefED committed:
improve json schema methods

YousefED committed:
fix build

YousefED committed:
WIP: schemas and selections

YousefED committed:
selections wip

YousefED committed:
update selections

YousefED committed:
wip selectionmarkers

YousefED committed:
drop core / react structure

YousefED committed:
ai menu

YousefED committed:
add comment

YousefED committed:
misc

matthewlipski committed:
Added `size` field to React suggestion menu items

YousefED committed:
Merge branch 'ai-block' of github.com:TypeCellOS/BlockNote into ai-block

YousefED committed:
basis for accept / reject menu

YousefED committed:
selection commands

matthewlipski committed:
Added `.env` file for API key and AI menu buttons for after an AI command is run

matthewlipski committed:
Merge remote-tracking branch 'origin/ai-block' into ai-block

matthewlipski committed:
Added loader to AI menu

matthewlipski committed:
Updated styles

YousefED committed:
wip

YousefED committed:
gitignore

YousefED committed:
Merge remote-tracking branch 'origin/main' into ai-block

YousefED committed:
move to xl-ai and add server

YousefED committed:
fix test

YousefED committed:
small merge fixes

YousefED committed:
Merge remote-tracking branch 'origin/main' into ai-block

YousefED committed:
update lock

YousefED committed:
update ai sdk

YousefED committed:
improve tests

YousefED committed:
model selector

YousefED committed:
add markdowndiff

YousefED committed:
wip

YousefED committed:
split and update tests

YousefED committed:
add list support

YousefED committed:
wip

YousefED committed:
Merge remote-tracking branch 'origin/main' into ai-block

YousefED committed:
merge

YousefED committed:
Merge remote-tracking branch 'origin/main' into ai-block

YousefED committed:
refactor

YousefED committed:
improve tests

YousefED committed:
update tests

YousefED committed:
wip

YousefED committed:
fixes

YousefED committed:
wip

YousefED committed:
fix tests

YousefED committed:
fix, before removing commented code

YousefED committed:
fix

YousefED committed:
fix tests

YousefED committed:
fix lint

YousefED committed:
add comments

YousefED committed:
update tests

YousefED committed:
fix

YousefED committed:
fix build

YousefED committed:
fix lock

YousefED committed:
wip

YousefED committed:
misc

YousefED committed:
remove unused files, fix basic toolbar ux

YousefED committed:
fix build

YousefED committed:
fix lint

YousefED committed:
fix build

YousefED committed:
fix build

YousefED committed:
fix start command

YousefED committed:
fix start

YousefED committed:
port

YousefED committed:
fix deploy

YousefED committed:
fix build

YousefED committed:
ai client env variables

YousefED committed:
add + fix test

YousefED committed:
add prepare scripts

YousefED committed:
test

YousefED committed:
albert

YousefED committed:
fix build

matthewlipski committed:
Made editor non-editable until user finishes LLM generation

matthewlipski committed:
Merge remote-tracking branch 'origin/ai-block' into ai-block

matthewlipski committed:
Implemented TODO LLM commands

YousefED committed:
disable fix

YousefED committed:
support selections

YousefED committed:
fix

YousefED committed:
fix

YousefED committed:
fix blank line processing with markdown

YousefED committed:
improve prompt

YousefED committed:
Merge remote-tracking branch 'origin/main' into ai-block

YousefED committed:
merge / undo button separation

YousefED committed:
revert some changes

YousefED committed:
Merge branch 'main' of github.com:TypeCellOS/BlockNote into ai-block

YousefED committed:
lockfile

YousefED committed:
remove unneeded files

YousefED committed:
Merge branch 'main' into ai-block

YousefED committed:
remove file

YousefED committed:
better streaming

YousefED committed:
refactor streams

YousefED committed:
update position

YousefED committed:
small ux fixes

YousefED committed:
remove file

YousefED committed:
fix build

YousefED committed:
fix some tests

YousefED committed:
fix some tests

YousefED committed:
fix tests

YousefED committed:
fix types, add test for errors

YousefED committed:
albert

YousefED committed:
wip

YousefED committed:
wip

YousefED committed:
add tests

YousefED committed:
misc

YousefED committed:
dontReplaceContentAtEnd

YousefED committed:
change snap paths

YousefED committed:
rewrite snapshot / test system

YousefED committed:
refactor

YousefED committed:
test improvements

YousefED committed:
markdownblocks :D

YousefED committed:
wip

YousefED committed:
upgrade ai package and add snapshot check

YousefED committed:
update json snaps

YousefED committed:
update msw snapshots

YousefED committed:
fix so me tests

YousefED committed:
update msw snaps

YousefED committed:
wip

YousefED committed:
wip html

YousefED committed:
wip

YousefED committed:
remove debug test

YousefED committed:
fix

YousefED committed:
Merge branch 'main' into ai-block

YousefED committed:
Merge remote-tracking branch 'origin/main' into ai-block

YousefED committed:
merge

YousefED committed:
html blocks fixes

YousefED committed:
fix build

YousefED committed:
revert tsconfig

YousefED committed:
Merge remote-tracking branch 'origin/main' into ai-block

YousefED committed:
update selections with new table logic

YousefED committed:
simplify some code re. selections

YousefED committed:
clean

YousefED committed:
wip

YousefED committed:
wip

YousefED committed:
wip

YousefED committed:
WIP: Zustand instead of context

nperez0111 commented on packages/xl-ai/src/AIExtension.ts:
```diff
@@ -0,0 +1,195 @@
+  return editor.extension(AIExtension);
+}
+
+// TODO: Revisit - this pattern might be a bit iffy
+// Make editor non-editable after calling the LLM, until the user accepts or
+// reverts the changes.
+// useEffect(() => {
+//   editor.isEditable = aiResponseStatus === "initial";
+// }, [aiResponseStatus, editor]);
```
Unrelated, but, this is why I prefer that LLM responses be injected as decorations instead of modifying document content.

Imagine a user refreshes their window, the LLM response should be dropped, but it actually modified the document. Just a thought

nperez0111 commented on packages/xl-ai/src/AIExtension.ts:
```diff
@@ -0,0 +1,195 @@
+    aiMenuResponseStatus: "initial",
+  }));
+
+  /**
+   * zustand design considerations
+   *
+   * - I decided to create a separate store for the options as they feel quite different from "internal state"
+   *   above. They're not so much "state", but more "configuration". I still think it's useful to have in a Zustand
+   *   store (instead of a just a plain object), because consumers might want to display the settings in the UI.
+   *   Case can also be made to make this a plain and leave the responsibility of displaying settings to the consumer
+   *   (and not have a store for this at all). TO DISCUSS
+   */
+  public readonly options: ReturnType<
+    ReturnType<typeof createStore<Required<GlobalLLMCallOptions>>>
+  >;
```
Yea, I think that configuration should just be a plain object, if the UI needs to be able to modify it, then the configur option should just be a default/initial value & the value is stored into the state store.

I'd imagine configuration like `codeBlock.supportedLanguages`, & I think it'd be complex to have that value be modifiable at runtime.

nperez0111 replied:
I'd even go as far as to say that configuration should be `T extends Readonly<Record<string, any>>`

YousefED replied:
not sure about this, let's discuss later (not zustand specific)

nperez0111 commented on packages/core/src/editor/BlockNoteEditor.ts:
```diff
@@ -698,6 +713,26 @@ export class BlockNoteEditor<
     this._tiptapEditor.dispatch(tr);
   };
 
+  // TO DISCUSS
+  /**
+   * Shorthand to get a typed extension from the editor, by
+   * just passing in the extension class.
+   *
+   * @param ext - The extension class to get
+   * @param key - optional, the key of the extension in the extensions object (defaults to the extension name)
+   * @returns The extension instance
+   */
+  public extension<T extends BlockNoteExtension>(
+    ext: { new (...args: any[]): T } & typeof BlockNoteExtension,
+    key = ext.name()
+  ): T {
+    const extension = this.extensions[key] as T;
+    if (!extension) {
+      throw new Error(`Extension ${key} not found`);
+    }
+    return extension;
+  }
```
I'm just gonna skip over any of the extension bit. I think there are better ways to do it, but not pertinent right now to discuss & should be separately discussed & written up

nperez0111 commented on packages/xl-ai/src/AIExtension.ts:
```diff
@@ -0,0 +1,195 @@
+    return "ai";
+  }
+
+  /**
+   * zustand design considerations
+   *
+   * - setters are not added in this store, but as class methods so we can have control over what to expose as API
+   * - con: if we'd expose the store externally, consumers could still call store.setState :/ (for this store it's not desired, for `options` below, it is)
+   *
+   */
+  public readonly store = createStore<AIPluginState>()((set) => ({
+    aiMenuBlockID: undefined,
+    aiMenuResponseStatus: "initial",
+  }));
```
Agree that setters being in the store would be weird, assuming that you have access to the class methods. If you didn't have access to the class methods, it would basically force all logic to be in zustand, which I think is the wrong way about things.

Being able to update state you don't own is pretty problematic 🙁  (if someone called store.setState), but I don't think we need to solve it or disable it.

YousefED replied:
> Being able to update state you don't own is pretty problematic 🙁 (if someone called store.setState), but I don't think we need to solve it or disable it.

agree - as long as everywhere else we follow the pattern that we expose separate setters, I think we should be good

nperez0111 commented on packages/xl-ai/src/AIExtension.ts:
```diff
@@ -0,0 +1,195 @@
+import { PromptOrMessages, llm } from "./api";
+import { LLMRequestOptions } from "./api/streamTool/callLLMWithStreamTools";
+
+type AIPluginState = {
+  aiMenuBlockID: string | undefined;
+  aiMenuResponseStatus: "initial" | "generating" | "error" | "done";
+};
```
What if this was a union, would setState be able to enforce this properly?

Like, what if it was:

```ts
type AIPluginState = {
   aiMenuBlockID: undefined;
   aiMenuResponseStatus: 'initial'
} | {
   aiMenuBlockID: string;
   aiMenuResponseStatus: "generating" | "error" | "done";
 };
```

YousefED replied:
good suggestion, see this commit: https://github.com/TypeCellOS/BlockNote/pull/980/commits/5a529cb49f4760e6ea56ea2179d9abfe824c7b8f

YousefED committed:
better type safety (zustand)

YousefED committed:
wip

YousefED committed:
wip

YousefED committed:
misc

YousefED committed:
fix

YousefED committed:
fix deps

YousefED committed:
fix build and slow stream

YousefED committed:
cleanup

YousefED committed:
wip

YousefED committed:
Merge remote-tracking branch 'origin/main' into ai-block

YousefED committed:
merge + fullslug for examples

YousefED committed:
cleanup

YousefED committed:
wip

YousefED committed:
improve tests + changeset

YousefED committed:
wip agent improvements

YousefED committed:
fix snaps

YousefED committed:
wip: support block type / attr changes, better table handling

YousefED committed:
updte tests

YousefED committed:
update prosemirror-transform

YousefED committed:
Merge remote-tracking branch 'origin/main' into ai-block

YousefED committed:
fix tests

YousefED committed:
tests async

YousefED committed:
clean tests

YousefED committed:
switch to mapper

YousefED committed:
refactor
